Realm/RLMObservation.mm

```diff
@@ -553,3 +555,69 @@ void RLMDidChange(std::vector<realm::BindingContext::ObserverState> const& obser
         static_cast<RLMObservationInfo *>(info)->didChange(RLMInvalidatedKey);
     }
 }
+
+KeyPath RLMKeyPathFromString(RLMRealm *realm, RLMSchema *schema, RLMObjectSchema *rlmObjectSchema, RLMClassInfo *info, NSString *keyPath) {
+    RLMProperty *property;
+    std::vector<std::pair<TableKey, ColKey>> keyPairs;
+
+    NSUInteger start = 0, length = keyPath.length, end = NSNotFound;
+    do {
+        end = [keyPath rangeOfString:@"." options:0 range:{start, length - start}].location;
+        NSString *propertyName = [keyPath substringWithRange:{start, end == NSNotFound ? length - start : end - start}];
+        property = rlmObjectSchema[propertyName];
+        RLMPrecondition(property, @"Invalid property name",
+                        @"Property '%@' not found in object of type '%@'",
+                        propertyName, rlmObjectSchema.className);
+
+        if (end != NSNotFound) {
+            RLMPrecondition(property.type == RLMPropertyTypeObject || property.type == RLMPropertyTypeLinkingObjects,
+                            @"Invalid value", @"Property '%@' is not a link in object of type '%@'",
+                            propertyName, rlmObjectSchema.className);
+            REALM_ASSERT(property.objectClassName);
+
+            TableKey tk = info->objectSchema->table_key;
+            ColKey ck;
+            if (property.type == RLMPropertyTypeObject) {
+                ck = info->tableColumn(property.columnName);
+                info = &realm->_info[property.objectClassName];
+            } else if (property.type == RLMPropertyTypeLinkingObjects) {
+                ck = info->computedTableColumn(property);
+                info = &realm->_info[property.objectClassName];
+            } else {
+                // This branch should never be reached. This case should have been caught by precondition above.
+                // How should an error be properly thrown?
+                RLMException(@"Precondition failure");
+            }
+
+            keyPairs.push_back(std::make_pair(tk, ck));
+            rlmObjectSchema = schema[property.objectClassName];
+        }
+
+        start = end + 1;
+    } while (end != NSNotFound);
+
+    TableKey tk = info->objectSchema->table_key;
+    ColKey ck;
+    if (property.type == RLMPropertyTypeLinkingObjects) {
+        ck = info->computedTableColumn(property);
+    } else {
+        ck = info->tableColumn(property.columnName);
+    }
+    keyPairs.push_back(std::make_pair(tk, ck));
+    return keyPairs;
+}
+
+// Some parameters ultimately not used.
+// TODO: clean up unused parameters
+KeyPathArray RLMKeyPathArrayFromStringArray(RLMRealm *realm,
+                                               RLMSchema *schema,
+                                               RLMObjectSchema *objectSchema,
+                                               RLMClassInfo *info,
+                                               NSArray<NSString *> *keyPaths) {
+    KeyPathArray rlmKeyPathArray;
```

Comment:
`keyPathArray` instead of `rlmKeyPathArray`?

Reply:
Yes. Will change.

Reply:
Resolved in [b8e9ee0](https://github.com/realm/realm-cocoa/pull/7249/commits/b8e9ee02afed35f06f77bd688886ad6b7aea627d)